File: reference/src/audit/logger.ts
/**
 * Append-only audit log with optional external sink.
 * Entries are frozen after creation to ensure immutability.
 * Hash chain provides tamper-evident integrity verification (PRD Section 11).
 */

import { createHash } from 'node:crypto';
import type { AuditEntry, AuditFilter, AuditLog, AuditSink, IntegrityResult } from './types.js';

let counter = 0;

function generateId(): string {
  return `audit-${Date.now()}-${++counter}`;
}

function matchesFilter(entry: AuditEntry, filter: AuditFilter): boolean {
  if (filter.actor && entry.actor !== filter.actor) return false;
  if (filter.action && entry.action !== filter.action) return false;
  if (filter.resource && entry.resource !== filter.resource) return false;
  if (filter.decision && entry.decision !== filter.decision) return false;
  if (filter.from && entry.timestamp < filter.from) return false;
  if (filter.to && entry.timestamp > filter.to) return false;
  return true;
}

/**
 * Compute a SHA-256 hash for an audit entry, chaining to the previous hash.
 * The hash covers all content fields plus the previousHash, excluding the hash field itself.
 */
export function computeEntryHash(entry: Omit<AuditEntry, 'hash'>, previousHash?: string): string {
  const payload = JSON.stringify({
    id: entry.id,
    timestamp: entry.timestamp,
    actor: entry.actor,
    action: entry.action,
    resource: entry.resource,
    policy: entry.policy,
    decision: entry.decision,
    details: entry.details,
    previousHash: previousHash ?? null,
  });
  return createHash('sha256').update(payload).digest('hex');
}

export function createAuditLog(sink?: AuditSink): AuditLog {
  const log: AuditEntry[] = [];

  return {
    record(partial: Omit<AuditEntry, 'id' | 'timestamp'> & { timestamp?: string }): AuditEntry {
      const prevHash = log.length > 0 ? log[log.length - 1].hash : undefined;

      const base = {
        id: generateId(),
        timestamp: partial.timestamp ?? new Date().toISOString(),
        actor: partial.actor,
        action: partial.action,
        resource: partial.resource,
        policy: partial.policy,
        decision: partial.decision,
        details: partial.details,
        previousHash: prevHash,
      };

      const hash = computeEntryHash(base, prevHash);

      const entry: AuditEntry = Object.freeze({
        ...base,
        hash,
      });

      log.push(entry);
      if (sink) {
        sink.write(entry);
      }
      return entry;
    },

    entries(): readonly AuditEntry[] {
      return log;
    },

    query(filter: AuditFilter): readonly AuditEntry[] {
      return log.filter((e) => matchesFilter(e, filter));
    },

    verifyIntegrity(): IntegrityResult {
      if (log.length === 0) return { valid: true };

      for (let i = 0; i < log.length; i++) {
        const entry = log[i];
        const expectedPrevHash = i > 0 ? log[i - 1].hash : undefined;

        // Verify chain link
        if (entry.previousHash !== expectedPrevHash) {
          return { valid: false, brokenAt: i };
        }

        // Recompute and verify entry hash
        const recomputed = computeEntryHash(entry, expectedPrevHash);
        if (entry.hash !== recomputed) {
          return { valid: false, brokenAt: i };
        }
      }

      return { valid: true };
    },
  };
}

File: sdk-go/audit/logger.go
package audit

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"sync"
	"time"
)

type auditLog struct {
	mu           sync.Mutex
	sinks        []AuditSink
	entries      []*AuditEntry
	previousHash string
	counter      int
}

// NewAuditLog creates a new audit log that writes to the given sinks.
func NewAuditLog(sinks ...AuditSink) AuditLog {
	return &auditLog{
		sinks:        sinks,
		previousHash: "genesis",
	}
}

func (a *auditLog) Log(action, resourceKind, resourceName, actor, outcome string, details map[string]interface{}) error {
	a.mu.Lock()
	defer a.mu.Unlock()

	a.counter++
	entry := &AuditEntry{
		ID:           fmt.Sprintf("audit-%d", a.counter),
		Timestamp:    time.Now().UTC().Format(time.RFC3339),
		Action:       action,
		ResourceKind: resourceKind,
		ResourceName: resourceName,
		Actor:        actor,
		Outcome:      outcome,
		Details:      details,
		PreviousHash: a.previousHash,
	}
	entry.Hash = ComputeEntryHash(entry)
	a.previousHash = entry.Hash
	a.entries = append(a.entries, entry)

	for _, sink := range a.sinks {
		if err := sink.Write(entry); err != nil {
			return fmt.Errorf("audit sink write failed: %w", err)
		}
	}
	return nil
}

func (a *auditLog) Query(filter *AuditFilter) ([]*AuditEntry, error) {
	a.mu.Lock()
	defer a.mu.Unlock()

	var results []*AuditEntry
	for _, e := range a.entries {
		if matchesFilter(e, filter) {
			results = append(results, e)
		}
	}
	if filter != nil && filter.Limit > 0 && len(results) > filter.Limit {
		results = results[len(results)-filter.Limit:]
	}
	return results, nil
}

func (a *auditLog) Verify() (bool, error) {
	a.mu.Lock()
	defer a.mu.Unlock()

	prev := "genesis"
	for _, e := range a.entries {
		if e.PreviousHash != prev {
			return false, nil
		}
		expected := ComputeEntryHash(e)
		if e.Hash != expected {
			return false, nil
		}
		prev = e.Hash
	}
	return true, nil
}

func matchesFilter(e *AuditEntry, f *AuditFilter) bool {
	if f == nil {
		return true
	}
	if f.Action != "" && e.Action != f.Action {
		return false
	}
	if f.ResourceKind != "" && e.ResourceKind != f.ResourceKind {
		return false
	}
	if f.ResourceName != "" && e.ResourceName != f.ResourceName {
		return false
	}
	if f.Actor != "" && e.Actor != f.Actor {
		return false
	}
	if f.Since != nil {
		ts, err := time.Parse(time.RFC3339, e.Timestamp)
		if err != nil || ts.Before(*f.Since) {
			return false
		}
	}
	if f.Until != nil {
		ts, err := time.Parse(time.RFC3339, e.Timestamp)
		if err != nil || ts.After(*f.Until) {
			return false
		}
	}
	return true
}

// ComputeEntryHash computes the SHA-256 hash for an audit entry (excluding the Hash field).
func ComputeEntryHash(entry *AuditEntry) string {
	data := map[string]interface{}{
		"id":           entry.ID,
		"timestamp":    entry.Timestamp,
		"action":       entry.Action,
		"resourceKind": entry.ResourceKind,
		"resourceName": entry.ResourceName,
		"actor":        entry.Actor,
		"outcome":      entry.Outcome,
		"details":      entry.Details,
		"previousHash": entry.PreviousHash,
	}
	raw, _ := json.Marshal(data)
	h := sha256.Sum256(raw)
	return hex.EncodeToString(h[:])
}
